executor/plan_recreator.go

```diff
@@ -0,0 +1,456 @@
+		if err := recordSets[0].Close(); err != nil {
+			return "", err
+		}
+	}
+
+	// Dump explain
+	if e.Analyze {
+		// Explain analyze
+		recordSets, err = e.Ctx.(sqlexec.SQLExecutor).Execute(context.TODO(), fmt.Sprintf("explain analyze %s", e.ExecStmt.Text()))
+		if err != nil {
+			return "", err
+		}
+	} else {
+		// Explain
+		recordSets, err = e.Ctx.(sqlexec.SQLExecutor).Execute(context.TODO(), fmt.Sprintf("explain %s", e.ExecStmt.Text()))
+		if err != nil {
+			return "", err
+		}
+	}
+	sRows, err = resultSetToStringSlice(context.Background(), recordSets[0])
+	if err != nil {
+		return "", err
+	}
+	fw, err := zw.Create("explain.txt")
+	if err != nil {
+		return "", err
+	}
+	for _, row := range sRows {
+		fmt.Fprintf(fw, "%s\n", strings.Join(row, "\t"))
+	}
+	if len(recordSets) > 0 {
+		if err := recordSets[0].Close(); err != nil {
+			return "", err
+		}
+	}
+	return hex.EncodeToString(token[:]), nil
+}
+
+func extractTableNames(ExecStmt ast.StmtNode, curDB string) (map[tableNamePair]struct{}, error) {
+	extractor := &tableNameExtractor{
+		curDB: curDB,
+		names: make(map[tableNamePair]struct{}),
+	}
+	ExecStmt.Accept(extractor)
+	return extractor.names, nil
+}
+
+func getStatsForTable(do *domain.Domain, pair tableNamePair) (*handle.JSONTable, error) {
+	is := do.InfoSchema()
+	h := do.StatsHandle()
+	tbl, err := is.TableByName(model.NewCIStr(pair.DBName), model.NewCIStr(pair.TableName))
+	if err != nil {
+		return nil, err
+	}
+	js, err := h.DumpStatsToJSON(pair.DBName, tbl.Meta(), nil)
+	return js, err
+}
+
+func getShowCreateTable(pair tableNamePair, zw *zip.Writer, ctx sessionctx.Context) error {
+	recordSets, err := ctx.(sqlexec.SQLExecutor).Execute(context.TODO(), fmt.Sprintf("show create table `%v`.`%v`", pair.DBName, pair.TableName))
+	if err != nil {
+		return err
+	}
+	sRows, err := resultSetToStringSlice(context.Background(), recordSets[0])
+	if err != nil {
+		return err
+	}
+	fw, err := zw.Create(fmt.Sprintf("schema/%v.%v.schema.txt", pair.DBName, pair.TableName))
+	if err != nil {
+		return err
+	}
+	for _, row := range sRows {
+		fmt.Fprintf(fw, "%s\n", strings.Join(row, "\t"))
+	}
+	if len(recordSets) > 0 {
+		if err := recordSets[0].Close(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func resultSetToStringSlice(ctx context.Context, rs sqlexec.RecordSet) ([][]string, error) {
+	rows, err := getRows4Test(ctx, rs)
+	if err != nil {
+		return nil, err
+	}
+	err = rs.Close()
+	if err != nil {
+		return nil, err
+	}
+	sRows := make([][]string, len(rows))
+	for i := range rows {
+		row := rows[i]
+		iRow := make([]string, row.Len())
+		for j := 0; j < row.Len(); j++ {
+			if row.IsNull(j) {
+				iRow[j] = "<nil>"
+			} else {
+				d := row.GetDatum(j, &rs.Fields()[j].Column.FieldType)
+				iRow[j], err = d.ToString()
+				if err != nil {
+					return nil, err
+				}
+			}
+		}
+		sRows[i] = iRow
+	}
+	return sRows, nil
+}
+
+func getRows4Test(ctx context.Context, rs sqlexec.RecordSet) ([]chunk.Row, error) {
```

Comment:
Can we use `DrainRecordSet()` directly, I think they're similar.

Reply:
It's only used for oom action.